src/_P088_Ping.ino

```diff
@@ -0,0 +1,258 @@
+struct P088_data_struct {
+  ip_addr_t destIPAddress[TASKS_MAX];
+  uint32_t idseq[TASKS_MAX];
+  struct raw_pcb *m_IcmpPCB = NULL;
+  uint8_t instances;
+};
+
+struct P088_data_struct *P088_data = NULL;
+
+boolean Plugin_088(byte function, struct EventStruct *event, String& string)
+{
+  boolean success = false;
+
+  switch (function)
+  {
+  case PLUGIN_DEVICE_ADD:
+  {
+    Device[++deviceCount].Number = PLUGIN_ID_088;
+    Device[deviceCount].Type = DEVICE_TYPE_DUMMY;
+    Device[deviceCount].VType = DEVICE_TYPE_SINGLE;
+    Device[deviceCount].Ports = 0;
+    Device[deviceCount].ValueCount = 1;
+    Device[deviceCount].PullUpOption = false;
+    Device[deviceCount].InverseLogicOption = false;
+    Device[deviceCount].DecimalsOnly = true;
+    Device[deviceCount].FormulaOption = false;
+    Device[deviceCount].SendDataOption = true;
+    Device[deviceCount].TimerOption = true;
+    break;
+  }
+
+  case PLUGIN_GET_DEVICENAME:
+  {
+    //return the device name
+    string = F(PLUGIN_NAME_088);
+    break;
+  }
+  case PLUGIN_GET_DEVICEVALUENAMES:
+  {
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_088));
+    break;
+  }
+
+  case PLUGIN_WEBFORM_LOAD:
+  {
+    char hostname[PLUGIN_088_HOSTNAME_SIZE];
+    LoadCustomTaskSettings(event->TaskIndex, (byte*)&hostname, PLUGIN_088_HOSTNAME_SIZE);
+    addFormTextBox(String(F("Hostname")), F("p088_ping_host"), hostname, PLUGIN_088_HOSTNAME_SIZE - 2);
+    success = true;
+    break;
+  }
+
+  case PLUGIN_WEBFORM_SAVE:
+  {
+    char hostname[PLUGIN_088_HOSTNAME_SIZE];
+    strncpy(hostname,  WebServer.arg(F("p088_ping_host")).c_str() , sizeof(hostname));
+    SaveCustomTaskSettings(event->TaskIndex, (byte*)&hostname, PLUGIN_088_HOSTNAME_SIZE);
+    success = true;
+    break;
+  }
+
+  case PLUGIN_INIT:
+  {
+    if (!P088_data) {
+      P088_data = new P088_data_struct();
+      P088_data->m_IcmpPCB = raw_new(IP_PROTO_ICMP);
+      raw_recv(P088_data->m_IcmpPCB, PingReceiver, NULL);
+      raw_bind(P088_data->m_IcmpPCB, IP_ADDR_ANY);
+      for (int i = 0; i < PLUGIN_088_MAX_INSTANCES; i++) {
+        memset(&P088_data->destIPAddress[i], 0x0, sizeof(P088_data->destIPAddress[0]));
+        memset(&P088_data->idseq[i], 0x0, sizeof(P088_data->idseq[0]));
+      }
+      P088_data->instances = 1;
+    } else {
+      P088_data->instances++;
+    }
+    UserVar[event->BaseVarIndex] = 0;
+    success = true;
+    break;
+  }
+
+  case PLUGIN_EXIT:
+  {
+    P088_data->instances--;
+    if (P088_data->instances == 0) {
+      raw_remove(P088_data->m_IcmpPCB);
+      delete P088_data;
+    }
+    break;
+  }
+
+  case PLUGIN_READ:
+  {
+    char hostname[PLUGIN_088_HOSTNAME_SIZE];
+    IPAddress ip;
+
+    // Does we have unanswered pings?
+    if (P088_data->destIPAddress[event->TaskIndex].addr != 0) {
+      UserVar[event->BaseVarIndex]++;
+    }
+
+    success = true;
+    if (!WiFiConnected()) {
+      break;
```

Comment:
Why set success to true, even if there is no WiFi connection?
Succes = true in the `PLUGIN_READ` means it will send its value to the connected controller(s) and generate an event to be processed in the rules.
I think it does make sense to set it only true if it changed the value. Or am I missing something here?

Reply:
Ah I get it now, also if you don't have WiFi, it is an indication something is wrong :)
Coffee starts to kick in now :)